proposals/accepted/051-ghc-base-libraries.rst

```diff
@@ -0,0 +1,330 @@
+  For a transitional period, ``base`` can continue to export the function, but with a deprecation warning saying something like:
+
+    This is going to disappear from base.
+    You probably don't want to use it at all.
+    But if you absolutely must, get it from ``ghc-internals``.
+
+- To expose a new function from ``ghc-internals`` requires that any functions on which it depends are also in ``ghc-internals`` (not base).
+  So we may need to move code from ``base`` to ``ghc-internals``, leaving a shim behind in base.
+  In practice, that may mean that quite a lot of code will move into ``ghc-internals`` quite quickly.
+  But that's fine: *it is just an implementation matter*: provided the modules, exports, and API of ``base`` are maintained, it is immaterial to clients (and hence to CLC) exactly *how* they are maintained.
+
+- This proposal is fully compatible with, and actively supports, the `CLC charter <https://github.com/haskell/core-libraries-committee#base-package>`__:
+
+    The primary responsibility of CLC is to manage API changes of ``base`` package.
+    The ownership of ``base`` belongs to GHC developers, and they can maintain it freely without CLC involvement as long as changes are invisible to clients.
+    Documentation changes normally fall under GHC developers purview, except significant ones (e.g., adding or changing type class laws).
+
+- It also supports GHC innovation, by
+
+  - allowing GHC freedom to change aspects of its implementation
+
+  - allowing the GHC Steering Committee to add new functions and types in ``ghc-experimental``.
+
+- One might wonder why GHC has three "internal" packages: ``ghc-internals``, ``ghc-bignum``, and ``ghc-prim``? Could they not be a single package? Answer: technically yes, but it helps to keep dependencies and responsibilities clear.
+  And it's purely an internal GHC matter; if the team wants to structure GHC's internals with three packages, or ten, that's up to them.
+
+Continuous integration
+======================
+
+A major difficulty is **knowing when the API of 'base' (as defined in Section 2) has changed.** A change requires CLC approval; but how do we know what commits (to ``base``, to ``ghc-internals``, to ``ghc-prim``) make such a change?
+
+In the past we have relied on best efforts; but with a bunch of volunteers, mistakes will be made.
+And mistakes can lead to a loss of trust.
+
+The solution is obvious: we need to automate.
+We therefore propose the following, as part of CI:
+
+1. Compile a good chunk of Hackage (around 500 packages) against base.
+   We already do this, and it is a huge help in reassuring ourselves that a change does not lead to accidental breakage.
+
+2. Test if any of the types (incl their kinds), functions (incl their types) and instances exposed by the ``base`` API are accidentally changed by a commit.
+   This is definitely going to happen, soon: @bgamari already has a prototype.
```

Comment:
This tool would be *amazing* to have for other libraries - I'd love to have something here so I could know for sure if a new release is versioned appropriately.

Reply:
I think this already exists as https://kowainik.github.io/posts/policeman-bristol - or is what you're looking for something else?

Reply:
`policeman` is a good approximation but when I examined it for use in GHC I found it had a few shortcomings:

 * it makes no attempt at comparing anything beyond the names of exports; that is, the types of bindings, exposed instances, `MINIMAL` declarations of classes,  and the like are not accounted for in the PVP assessment
 * there isn't a clear way to handle packages with platform-dependent exports; `base` is (unfortunately) such a package
 * it relies on `.hie` files, which we currently don't produce in the GHC build